### PR TITLE
perf(musa): bound DSV4 long-prefill indexer logits for 128k

### PR DESCRIPTION
**Pull request by Joey-gvwal**
## Summary

The production path for DeepSeek-V4 on MUSA is 128K as the primary path, with 256K as the upper limit, while the product also needs to support 1M context. In upstream, when the compressed key length is >4096, the sparse indexer directly skips the materialized paged-MQA + native `sparse_indexer_topk` path. As a result, long-context prefill falls back to a slower path, which is the issue this PR aims to address.

## Modification

| file | changed |
|---|---|
| `vllm_musa/patches/series/0166-MUSA-bound-DSV4-long-prefill-indexer-logits.patch` | patch `vllm/model_executor/layers/sparse_attn_indexer.py`  build-time |
| `vllm_musa/optimization_contract/policy.py` | export logits  / Gloo gate |
| `vllm_musa/optimization_contract/__init__.py` | re-export query |
| `tests/test_deepseek_v4_long_prefill_indexer_patch.py` | contract test |
| `vllm_musa/patches/series/README.md` | series readme |

### Start serve

```bash
export PYTHONUNBUFFERED=1
export TRANSFORMERS_OFFLINE=1
export HF_HUB_OFFLINE=1
export VLLM_WORKER_MULTIPROC_METHOD=spawn

python3 -m vllm.entrypoints.cli.main serve /ipfs/test/models/DeepSeek-V4-Flash-0731-FP8-mt \
  --trust-remote-code \
  --tensor-parallel-size 8 \
  --served-model-name DeepSeek-V4-Flash \
  --port 28500 \
  --max-model-len 1048576 \
  --max-num-seqs 1 \
  --max-num-batched-tokens 4096 \
  --gpu-memory-utilization 0.95 \
  --kv-cache-dtype fp8 \
  --enable-prefix-caching \
  --attention-backend FLASHMLA \
  --enable-chunked-prefill \
  --long-prefill-token-threshold 4096 \
  --speculative-config '{"method":"dspark","num_speculative_tokens":4}' \
  --compilation-config '{"mode":"NONE","cudagraph_mode":"FULL_DECODE_ONLY","cudagraph_capture_sizes":[1,2,3,4,5,8,16,24,32,40,48,56,64]}' \
  --async-scheduling \
  --hf-overrides '{"expert_dtype":"fp8"}'
```

### Bench

```bash
export TRANSFORMERS_OFFLINE=1
export HF_HUB_OFFLINE=1

python3 -m sglang.bench_serving \
  --backend vllm-chat \
  --host 127.0.0.1 \
  --port 28500 \
  --dataset-name random \
  --dataset-path /dataset/ShareGPT_V3_unfiltered_cleaned_split.json \
  --random-input-len 131072 \
  --random-output-len 1 \
  --random-range-ratio 1 \
  --output-details \
  --warmup-requests 1 \
  --seed 1 \
  --burstiness 102 \
  --request-rate 1 \
  --num-prompts 1 \
  --max-concurrency 1 \
  --tokenizer /ipfs/test/models/DeepSeek-V4-Flash-0731-FP8-mt \
  --model DeepSeek-V4-Flash 
```

### Result

| r | Mean TTFT (ms) |
|---|---|
| r1 | 1507（warmup 打 MHC JIT） |
| r2 | 708 |
| r3 | 724 |
| r4 | 705 |
| r5 | 707 |
| r6 | 707 |
| **r2–r6 mean** | **710 ms** |

### CODE REVIEW

File: vllm_musa/patches/series/0170-MUSA-bound-DSV4-long-prefill-indexer-logits.patch [hunk 1]
```diff
@@ -0,0 +1,404 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
```
Comment (yeahdongcn) [inline vllm_musa/patches/series/0170-MUSA-bound-DSV4-long-prefill-indexer-logits.patch:1]: Please rebase, 169 is taken.

Comment (Joey-gvwal) [reply vllm_musa/patches/series/0170-MUSA-bound-DSV4-long-prefill-indexer-logits.patch:1]: Done. I’ve rebased the changes and updated the patch numbering.
